pre_commit/languages/download.py

```diff
@@ -0,0 +1,318 @@
+run_hook = lang_base.basic_run_hook
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+class Platform:
+    OS: Final[Mapping[str, str]] = MappingProxyType({
+        'Linux': 'linux',
+        'Darwin': 'darwin',
+        'Windows': 'windows',
+        'DragonFly': 'dragonfly',
+        'FreeBSD': 'freebsd',
+    })
+    CPU: Final[Mapping[str, str]] = MappingProxyType({
+        'aarch64': 'arm64',
+        'aarch64_be': 'arm64be',
+        'arm': 'arm',
+        'i386': '386',
+        'i686': '386',
+        'x86_64': 'amd64',
+        'AMD64': 'amd64',
+        'ppc': 'ppc',
+        'ppc64': 'ppc64',
+        'ppc64le': 'ppc64le',
+    })
+
+    def __init__(self, value: str) -> None:
+        self.__value = value
+        os, cpu = self.parts
+        if os not in self.OS.values():
+            raise ValueError(f"invalid operating system `{os}`, \
+valid values are: {','.join(self.OS.values())}")
+        if cpu not in self.CPU.values():
+            raise ValueError(f"invalid CPU `{cpu}`, \
+valid values are: {','.join(self.CPU.values())}")
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> str:
+        os, _ = self.parts
+        return os
+
+    @property
+    def cpu(self) -> str:
+        _, cpu = self.parts
+        return cpu
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @classmethod
+    def host(cls: type[Platform]) -> Platform:
+        os = cls.OS[platform.system()]
+        cpu = cls.CPU[platform.machine()]
+        return cls(f'{os}/{cpu}')
+
+    def __str__(self) -> str:
+        return f'{self.os}/{self.cpu}'
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Platform):
+            return False
+        return other.os == self.os and other.cpu == self.cpu
+
+
+class ChecksumMismatchError(ValueError):
+    def __init__(
+        self, *, expected: SRI, actual: SRI,
+        message: str = 'checksum mismatch',
+    ):
+        self.__expected = expected
+        self.__actual = actual
+        self.__message = message
+
+    @property
+    def expected(self) -> SRI:
+        return self.__expected
+
+    @property
+    def actual(self) -> SRI:
+        return self.__actual
+
+    @property
+    def message(self) -> str:
+        return self.__message
+
+    def __str__(self) -> str:
+        return f'''{self.__message}:
+ - expected: {self.expected}
+ - actual  : {self.actual}
+'''
+
+
+class HasBinaryRead(Protocol):
+    def read(self, __size: int = -1) -> bytes | None: ...
```

Comment:
The other day I learnt that `_typeshed` has [various stubs](https://github.com/python/typeshed/blob/863423fbfe5b2e488f8a3e844ca853a4d6fda023/stdlib/_typeshed/__init__.pyi#L224) that can be reused.

One can do the following:

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from _typeshed import SupportsRead # coverage: ignore
```

The use `SupportsRead[bytes]` later.

You do not have to change this but thought I'd knowledge share :)

Reply:
What does ``# coverage: ignore`` do? It seems to work fine without it.